### ISSUE TITLE
Include Program Build Options as Part of the Hash

### ISSUE THREAD

**Issue by trbauer** [intel/opencl-intercept-layer#27]
## Observed Behavior

Currently the hashes used in `DevicePerformanceTimeHashTracking` do not include the program build options its calculation.  This makes it impossible to distinguish between programs that differ only in build options. Many OpenCL projects heavily (ab)use the C preprocessor (thus build options) and different options can radically change the program behavior based on those. 

Two examples:

    #ifdef USE_IMAGES
       pix = read_imageui(...);
    #else // use buffers
       pix = buffer[...];
    #endif

Another example is programs that hardcode the workgroup size in for things like loop unrolling.

     local tile[TILE_SIZE];
     for (int i = 0; i < TILE_SIZE; i++)
         ...

Presumably we want the program to be distinguished based on the `TILE_SIZE` chosen. 

## Desired Behavior

This issue proposes we simply include the build options as part of the hash. In the intercept log, the hash will be part of the same .cl file dumped and the user can track down the options the program was built with as needed.

## Steps to Reproduce

Enable `DevicePerformanceTimeHashTracking` and `DevicePerformanceTiming` and build two programs using the same input and different build options and invoke kernels from both programs.  In the `DevicePerformance` output file (`clintercept_report.txt`) note that the hash part for each program does not differ (the program and build count still differ.

**Comment by bashbaug** [comment]
There are actually three parts to the "hash" today.  Here's a snippet from a device performance timing report:

````
                       Function Name,  Calls, ...
GenerateJuliaSet(0000_3F40E1CD_0001),      4, ...
````

1. The first number is the program number, which is optional, but useful in many contexts when programs are created and built deterministically.  It gets assigned when the program is created.

1. The second number is the actual hash value, computed from the program source, IL, or input binary.  It also gets assigned when the program is created.

1. The third number is the compile count, since the same program can (theoretically, at least) be built multiple times with different options.  The compile count is initialized to zero when the program is created, and incremented each time the program is built.

Note that most parts of the "hash" are determined when the program is created, but that the program build options are only known when the program is built.  In particular, the actual hash value is computed when the program is created, and is used to determine the file name when dumping the program source, IL, or input binary, etc.

We could move much of the functionality that's currently performed when the program is created to when the program is built, but perhaps there's a better way: could we hash the build options separately, and include it in things like the device performance time report, but not necessarily in the program source file name?  I think this allows easy differentiation between two program objects that were built from the same source but different build options with minimal changes to the overall flow.  It may even allow removal of the "compile count", but I'm not completely convinced of that yet.

Example:

````
                                Function Name,  Calls, ...
GenerateJuliaSet(0000_3F40E1CD_ABCD1234_0001),      4, ...
                 ^     ^       ^        ^
 program number -+     |       |        |
 program source hash --+       |        |
          build options hash --+        |
     compile count, possibly redundant -+
````


Let me know if this will work for your needs - thanks!

**Comment by trbauer** [comment]
So with `clBuildProgram` what does the `cl_program` refer to when I build the program K times?  What does that even mean?  Could I call `clCreateKernel` on the first and then stomp the program?  

I favor value-based equality because it's more mathematically consistent to me. And further, I can't conjure up any example where we need build indexes (maybe there are?). If the hash is good enough to be unique I actually prefer that we only have one file if the same source inputs are used.

   CLI_PROG_3F40E1CD.cl

It reduces source clutter and makes it immediately how many .cl files there really are. For example, many filter benchmarks use the same source file over and over and just tweak the build options (e.g. filter radius). At a glance the reader can see that there's really one one unique input and just a mess of options.

If we're worried about disproving a hash collision on each build, source files are small enough that it's probably reasonable to do an equality check. Even the biggest programs are less than 100k. Remember, the compiler is going to do many order of magnitudes more work on that same 100k. And it's understood that program building calls can be expensive already. Moreover, we can omit that time out of the host time logging. If we have a true collision; use some sort of deterministic probing strategy.... 

Moreover, we can play the same game with build options. E.g. CLI_OPTS_ABCD1234.txt.  Two build strings are equal if and only if they are in the same file.

Another option is to disable/enable consideration of program index and build index via a control?

In the end it won't ruin my life either way.

In the device output I actually want kernels to be 

   (kernel_name x program_hash x build_opts_hash)

In other words, if one builds the same program twice with the same source and build options, I would want that to be in one single row. The bits executing will be the same. (Presuming the compiler is deterministic, which is pretty safe.) This gives us a clean definition via structural equality rather than nominal (representing unique build instances as part of the identity).

**Comment by trbauer** [comment]
So with `clBuildProgram` what does the `cl_program` refer to when I build the program K times?  What does that even mean?  Could I call `clCreateKernel` on the first and then stomp the program?  

I favor value-based equality because it's more mathematically consistent to me. And further, I can't conjure up any example where we need build indexes (maybe there are?). If the hash is good enough to be unique I actually prefer that we only have one file if the same source inputs are used.

   CLI_PROG_3F40E1CD.cl

It reduces source clutter and makes it immediately how many .cl files there really are. For example, many filter benchmarks use the same source file over and over and just tweak the build options (e.g. filter radius). At a glance the reader can see that there's really one one unique input and just a mess of options.

If we're worried about disproving a hash collision on each build, source files are small enough that it's probably reasonable to do an equality check. Even the biggest programs are less than 100k. Remember, the compiler is going to do many order of magnitudes more work on that same 100k. And it's understood that program building calls can be expensive already. Moreover, we can omit that time out of the host time logging. If we have a true collision; use some sort of deterministic probing strategy.... 

Moreover, we can play the same game with build options. E.g. CLI_OPTS_ABCD1234.txt.  Two build strings are equal if and only if they are in the same file.

Another option is to disable/enable consideration of program index and build index via a control?

In the end it won't ruin my life either way.

In the device output I actually want kernels to be 

   (kernel_name x program_hash x build_opts_hash)

In other words, if one builds the same program twice with the same source and build options, I would want that to be in one single row. The bits executing will be the same. (Presuming the compiler is deterministic, which is pretty safe.) This gives us a clean definition via structural equality rather than nominal (representing unique build instances as part of the identity).

**Comment by trbauer** [comment]
So with `clBuildProgram` what does the `cl_program` refer to when I build the program K times?  What does that even mean?  Could I call `clCreateKernel` on the first and then stomp the program?  

I favor value-based equality because it's more mathematically consistent to me. And further, I can't conjure up any example where we need build indexes (maybe there are?). If the hash is good enough to be unique I actually prefer that we only have one file if the same source inputs are used.

    CLI_PROG_3F40E1CD.cl

It reduces source clutter and makes it immediately how many .cl files there really are. For example, many filter benchmarks use the same source file over and over and just tweak the build options (e.g. filter radius). At a glance the reader can see that there's really one one unique input and just a mess of options.

If we're worried about disproving a hash collision on each build, source files are small enough that it's probably reasonable to do an equality check. Even the biggest programs are less than 100k. Remember, the compiler is going to do many order of magnitudes more work on that same 100k. And it's understood that program building calls can be expensive already. Moreover, we can omit that time out of the host time logging. If we have a true collision; use some sort of deterministic probing strategy.... 

Moreover, we can play the same game with build options. E.g. CLI_OPTS_ABCD1234.txt.  Two build strings are equal if and only if they are in the same file.

Another option is to disable/enable consideration of program index and build index via a control?

In the end it won't ruin my life either way.

In the device output I actually want kernels to be 

   (kernel_name x program_hash x build_opts_hash)

In other words, if one builds the same program twice with the same source and build options, I would want that to be in one single row. The bits executing will be the same. (Presuming the compiler is deterministic, which is pretty safe.) This gives us a clean definition via structural equality rather than nominal (representing unique build instances as part of the identity).

**Comment by bashbaug** [comment]
The changes added by PR #47 are intended to address this issue.  Can you please check and see if anything more is needed or if this issue may be closed?  Thanks!

**Comment by bashbaug** [comment]
I'm going to close this issue since I think everything is fixed.  If something is still missing, feel free to reopen it or to add a new issue.  Thanks!
